bundles/org.openhab.binding.chromecast/src/main/java/org/openhab/binding/chromecast/internal/ChromecastAudioSink.java

```diff
@@ -30,52 +37,103 @@
 
+    private static final Set<AudioFormat> SUPPORTED_FORMATS = Set.of(AudioFormat.MP3, AudioFormat.WAV);
+    private static final Set<Class<? extends AudioStream>> SUPPORTED_STREAMS = Set.of(AudioStream.class);
+
     private static final String MIME_TYPE_AUDIO_WAV = "audio/wav";
     private static final String MIME_TYPE_AUDIO_MPEG = "audio/mpeg";
 
-    private final ChromecastCommander commander;
+    private final ChromecastHandler handler;
     private final AudioHTTPServer audioHTTPServer;
     private final @Nullable String callbackUrl;
 
-    public ChromecastAudioSink(ChromecastCommander commander, AudioHTTPServer audioHTTPServer,
+    public ChromecastAudioSink(ChromecastHandler handler, AudioHTTPServer audioHTTPServer,
             @Nullable String callbackUrl) {
-        this.commander = commander;
+        this.handler = handler;
         this.audioHTTPServer = audioHTTPServer;
         this.callbackUrl = callbackUrl;
     }
 
-    public void process(@Nullable AudioStream audioStream) throws UnsupportedAudioFormatException {
+    @Override
+    public String getId() {
+        return handler.getThing().getUID().toString();
+    }
+
+    @Override
+    public @Nullable String getLabel(@Nullable Locale locale) {
+        return handler.getThing().getLabel();
+    }
+
+    @Override
+    public void processAsynchronously(@Nullable AudioStream audioStream) throws UnsupportedAudioFormatException {
         if (audioStream == null) {
             // in case the audioStream is null, this should be interpreted as a request to end any currently playing
             // stream.
             logger.trace("Stop currently playing stream.");
-            commander.handleCloseApp(OnOffType.ON);
+            handler.stop();
         } else {
             final String url;
             if (audioStream instanceof URLAudioStream) {
                 // it is an external URL, the speaker can access it itself and play it.
                 URLAudioStream urlAudioStream = (URLAudioStream) audioStream;
                 url = urlAudioStream.getURL();
+                tryClose(audioStream);
             } else {
                 if (callbackUrl != null) {
                     // we serve it on our own HTTP server
                     String relativeUrl;
-                    if (audioStream instanceof FixedLengthAudioStream) {
-                        relativeUrl = audioHTTPServer.serve((FixedLengthAudioStream) audioStream, 10);
-                    } else {
-                        relativeUrl = audioHTTPServer.serve(audioStream);
+                    try {
+                        StreamServed streamServed = audioHTTPServer.serve(audioStream, 10, true);
+                        relativeUrl = streamServed.url();
+                        // we have to run the delayed task when the server has completely played the stream
+                        streamServed.playEnd().thenRun(() -> this.playbackFinished(audioStream));
```

Comment:
@dalgwen : I assume that you also have to call `handler.stop();` when the playback ends ?

Reply:
AFAIK, the handler.stop() call we do line 76 is only because a null stream means we have to immediately stop playing.
Based on the previous code, it seems not needed anywhere else ?

Reply:
You mean Chromecast is stopping properly by itself without this call ?

With the Freebox audio sink, it was needed to do it to exit from the player app. But of course, it depends on every audio sink.

If not needed for the chromecast, that is fine.

Reply:
It depends on what "properly" means.
The player doesn't close, and it stays on the chromecast screen. The behavior has not changed in this pull request, it was like this before.

I just tested your suggestion : if I call the handler.stop() method when the playback is finished, it closes the chromecast screen after. It is IMHO a better behaviour.

So, we have here an opportunity. We are now able to guess more or less when play ends, and close the chromecast screen properly. BUT, as you already see, sound duration computation is not perfect. And so if we do, we have the risk of closing it before the sound is fully played.
For OGG, AAC sound in particular, it doesn't work at all. So I think we should let it that way, for now. Another option is a conditional close on sound codec. WDYT ?

Reply:
Your argument is good. Let's keep it unchanged for the chromecast sink.